### ISSUE TITLE
Provide internet access through wireless interface

### ISSUE THREAD

**Issue by alternatexp** [wifiphisher/wifiphisher#511]
Right now we're only able to use ethernet adapters, Wifiphisher would be one **BEAST** of a tool if it allowed internet access through a wireless interface (should (just) be a matter of iptables routing).

Phishing on the go, wherever you are!!

**Comment by sophron** [comment]
This is the continue of #270. 

The solution is not trivial. We need to have a managed wireless interface (that is connected on the Internet) and another one to pass to `hostapd` (which will turn it to `AP` mode). The thing is that `NetworkManager` will interfere with these interfaces if we leave him alive, but we can't kill him either if we want to provide Internet access from a WLAN.  There are a few hacks that we may end up doing to tell `NetworkManager` not to interfere with our interfaces.

**Comment by selli69** [comment]
Hmmm.... I have no network-manager (not installed either) running on my system and wiphisher runs like a charm (if cabled). In my opinion nm is a pain in the ass regarding wifi configuration. I dont know why anyone would have this crap running, when dealing with stuff like wiphisher. My remote raspberry connects automatically at startup via wifi to my AP due to configuration in /etc/network/interfaces, all without network manager.

**Comment by alternatexp** [comment]
@selli69 I fully agree with you. I find it unbelievable that all Kali distributions come with NetworkManager preinstalled. It is an absolute pain and will only cause you trouble.

**Comment by laozi999** [comment]
@sophron one way would be to kill NetworkManager on startup, then add a menu for the chosen `internetinterface` to select the AP with internet access and connect to it using dhclient.

**Comment by laozi999** [comment]
And the following commands to automate the process:

`--internetinterface "wlan3" --internetbssid "aa:bb:cc:dd:ee:ff" --internetkey "wifipassword"`
or
`-iI "wlan3" -iB "aa:bb:cc:dd:ee:ff" -iK "wifipassword"`

**Comment by selli69** [comment]
@sophoron @laozi999 

I think a menu for AP-Connection is pretty useless, because if you want to use wifiphisher with wireless upstream, you are already connected via wifi and have no extra cabled connection to do so. Also the command line options for AP-connect are useless in this scope.

Please consider having not only wireless interfaces for upstream. Also 3G/4G interfaces should be useable. Establishing the uplink is absolutely out of scope for a phishing tool. Not the same OSI Layer!

Simply make any interface type useable in wifipfisher.

**Comment by laozi999** [comment]
@selli69 this would not be the case if the person has connected to the wifi via NetworkManager.

I think it's best to ditch all the third party apps and set everything up fresh through Wifiphisher.

**Comment by anakin1028** [comment]
@sophron why don't we use the `wpa_supplicant` to connect to WLAN instead of `nm`?
BTW can I sidetrack ask what's the main purpose of the `-iI` option for now? 
From the code, I think the only function now is that the victim can get to internet from out fakeap. If users want any credential information, they have to sniff the ethernet interface [i.e the interface connect to WLAN]?

**Comment by sophron** [comment]
> BTW can I sidetrack ask what's the main purpose of the -iI option for now?
From the code, I think the only function now is that the victim can get to internet from out fakeap. If users want any credential information, they have to sniff the ethernet interface [i.e the interface connect to WLAN]?

Yes, a user may use `-iI` with a wired interface and then sniff that interface for sensitive data.

>why don't we use the wpa_supplicant to connect to WLAN instead of nm?

Killing NetworkManager and using wpa_supplicant is an option. The problem that I see is that the user will lose his connection (when we kill NetworkManager) and we will have to provide all functionality for connecting to the WLAN with wpa_supplicant (the user should provide ESSID / PSK). We will need a wpa_supplicant wrapper for this.

It would be nicer if the user can connect on his own and then simply pass the interface that has Internet connectivity to the script. 

I think the most elegant way is to instruct NetworkManager not to interfere with our interfaces. We can do that by leveraging NetworkManagers D-Bus interface. There is already a project that does this and we can probably include: https://github.com/seveas/python-networkmanager

**Comment by anakin1028** [comment]
@sophron you are right I have missed the part that user cannot lose the connection.
I try to use `wpa_supplicant` to get internet first and indeed `airmon-ng` will kill the connection.
I may study the module you post first :)

**Comment by sophron** [comment]
> @sophron you are right I have missed the part that user cannot lose the connection.
I try to use wpa_supplicant to get internet first and indeed airmon-ng will kill the connection.

If we find a way to talk with NetworkManager to instruct it not to mess with our interfaces, we can potentially remove `airmon-ng` from our script.

>I may study the module you post first :)

Sure! A few months ago I was able to achieve the desired result, by using the <a href="http://support.qacafe.com/knowledge-base/how-do-i-prevent-network-manager-from-controlling-an-interface/">KEYFILE</a> method. This is a hackish way though (writing to NetworkManager.conf) and I'm not sure it will work across different distros. Doing it via D-Bus would be a neat way.

**Comment by anakin1028** [comment]
Yes, I have seen the KEYFILE method too. 

And also I have done some research about d-bus today, though the following link shows that the adaptor may set to `un-managed`:

[Managed Property](https://developer.gnome.org/NetworkManager/unstable/gdbus-org.freedesktop.NetworkManager.Device.html#gdbus-property-org-freedesktop-NetworkManager-Device.Managed)

from my experiment this property seems to be un-settable.
The following is my test code just for reference, when I set the Managed property to `False`, exception raises. I don't know whether @sophron or @blackHatMonkey have D-Bus related experiences?

```python
import dbus
APP_PATH = 'org.freedesktop.NetworkManager'
NW_MANAGER_OBJ_PATH = '/org/freedesktop/NetworkManager'
#interface that nm implements
NW_MANAGER_INTERFACE_PATH = 'org.freedesktop.NetworkManager'

#interface that device object implements
DEV_INTERFACE_PATH = 'org.freedesktop.NetworkManager.Device'

bus = dbus.SystemBus()

'''
Get the wireless objects first
'''
#get nm proxy object
nm_proxy = bus.get_object(APP_PATH, NW_MANAGER_OBJ_PATH)
#from the proxy object get the nm interface object
nm = dbus.Interface(nm_proxy, dbus_interface=NW_MANAGER_INTERFACE_PATH)
#we can get the object paths for the devices
devs = nm.GetDevices()
print devs

for dev_obj_path in devs:
    dev_proxy = bus.get_object(APP_PATH, dev_obj_path)
    #since device interface doesn't implement methods to access properties
    #use org.freedesktop.Dbus.Properties instead
    dev = dbus.Interface(dev_proxy, dbus_interface='org.freedesktop.DBus.Properties')
    if dev.Get(DEV_INTERFACE_PATH, 'Interface') == 'wlan0':
        print 'managed for wlan0:', dev.Get(DEV_INTERFACE_PATH, 'Managed')
        dev.Set(DEV_INTERFACE_PATH, 'Managed', False)
```
here is a good introdution to D-Bus : [D-Bus](https://zignar.net/2014/09/08/getting-started-with-dbus-python-systemd/)

**Comment by blackHatMonkey** [comment]
@sophron @anakin1028 The problem with this approach is that it only works for `NetworkManager` and not any other managers. Now I know that our target operation system is `kali` and `kali` uses `NetworkManger` but it doesn't mean that should try to "hack" something together. Now personally I don't like this feature as it is incomplete but that's beside the point. I think it would be the best case if we let the user handle this and just provide us with the interface connected to the internet. On a side note I think it would be better to remove the `airmon-ng` call to kill a bunch of services and on exit it leaves the system unusable to say the least. If we are closing a program or a service we should make sure to run it back or better yet don't touch unless we absolutely have to. That is just my two cents :smile: .

**Comment by anakin1028** [comment]
@blackHatMonkey I agree with your points, but the problem is when the `NetworkManager` is running, it will interfere our hostapd and maybe deauth interfaces, so that hostapd cannot be correctly launched.
If we don't implement this feature then users must specify the interfaces used as fakeap and jamming interfaces in the KEYFILE mentioned by @sophron, so that NetworkManager will not interfere with these interfaces.

**Comment by anakin1028** [comment]
I think I will read the code of NetworkManafer and see why it cannot accept setting the Managed property. 
It is still valuable to do this function in my view, since doing this we don't need to violently kill all the Network manager process[wifiphisher now doing this] and users don't need to configure their KEYFILE.

**Comment by anakin1028** [comment]
@sophron @blackHatMonkey 
It looks like the `Managed` property cannot be dynamically set as the following link:
[Managed Property unsettable](https://ubuntuforums.org/archive/index.php/t-2098623.html)

I think we can remove the code of `airmon-ng` since it will kill the wireless interface used to connect to the internet. After that we may have two approaches now:

1. After adapter selection done, I can write the KEYFILE and set the mac addresses of AP and deauth interfaces as `un-managed` [may have risk for different Linux distribution].

or

2. We only  do the sanity-check for the manged property of the network interfaces, if it is `managed` we raise exception.

This is what I can think now, how do you guys think? Thanks.

**Comment by JavaRockstar** [comment]
I was looking for something like this for a while and I came across something called airgeddon the problem i was facing was when network manager put the access points down the internet would go down too thought id mention incase anyone could get an idea of this I tried many different ways but this was the best I could find that keeps the internet going. I'd love to add a wifi interface to supply internet to wifiphisher as with newer devices that require a sim card ignore the wifi connection if it has no internet to give and the fake page does not show if the wireless ap had internet the phone or device chooses the wireless connection first.

Thanks device I've been testing on is a Android Samsung S7 Edge I know this is a bit of a challenge but I think it could be a feature that could need added by the way devices are starting to handle connections basically on my device it will say connected to AP but then say No Internet I like this approach without Internet but maybes a fake internet indicator would work by tricking the device into thinking it has Internet when really it hasint. I know wifiphisher wasint designed for this but it would be nice to add a repeater section to turn wifiphisher into a repeater to extend internet connection or host a captive portal with internet I know this has been covered and I dont want to come across as bringing back an old topic up although of current times it would be nice to add support for Internet and SSLStrip2 support as sites such as Google give by a warning message on the first connect.

I have worked with IPTables before and isc-dchp lighttpd before and found airgeddon to do a nice job but it does not support phishing and was thinking to my self Wifiphisher could be a lot better with these options is there any plans for this feature thanks also cheers for the advice on the Phishing Scenarios BlackHatMonkey.
